L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc

```diff
@@ -1,52 +1,73 @@
+    edm::LogInfo("HGCclusterParameters") << "C2d seeding Thr: " << seedThr_ ; 
+    edm::LogInfo("HGCclusterParameters") << "C2d clustering Thr: " << tcThr_ ; 
+
+    bool isSeed[trgcells_.size()];
+
+    /* seeding the TCs */
+    int itc=0;
+    for( l1t::HGCalTriggerCellBxCollection::const_iterator tc = trgcells_.begin(); tc != trgcells_.end(); ++tc,++itc )
+        isSeed[itc] = ( tc->hwPt() > seedThr_) ? true : false;
+
+    /* clustering the TCs */
+    std::vector<l1t::HGCalCluster> clustersTmp;
+
+    itc=0;
+    for( l1t::HGCalTriggerCellBxCollection::const_iterator tc = trgcells_.begin(); tc != trgcells_.end(); ++tc,++itc ){
+
+        if( tc->hwPt() < tcThr_ )
+            continue;
+
+        /* searching for TC near the center of the cluster  */
+        int iclu=0;
+        vector<int> tcPertinentClusters; 
+        for( std::vector<l1t::HGCalCluster>::iterator clu = clustersTmp.begin(); clu != clustersTmp.end(); ++clu,++iclu )
+            if( clu->isPertinent(*tc, dr_) )
+                tcPertinentClusters.push_back(iclu);
+
+        if( tcPertinentClusters.size() == 0 && isSeed[itc] ){
+            //edm::PtrVector<l1t::HGCalTriggerCell> coll;
+            //clustersTCcollections_.push_back( coll );
+            l1t::HGCalCluster obj( *tc );
+            clustersTmp.push_back( obj );
+        }
+        else if ( tcPertinentClusters.size() > 0 ){
+         
+            uint minDist = 300;
+            uint targetClu = 0;
                         
-                if(trgCellLayer<28){
-                    C2d_hwPtEm+=tc->hwPt();
-                }else if(trgCellLayer>=28){
-                    C2d_hwPtHad+=tc->hwPt();
+            for( std::vector<int>::const_iterator iclu = tcPertinentClusters.begin(); iclu != tcPertinentClusters.end(); ++iclu ){
+                double d = clustersTmp.at( *iclu).dist(*tc);
+                if( d < minDist ){
+                    minDist = d;
+                    targetClu = *iclu;
                 }
-                               
-                C2d_pt += tc->pt();                        
-                protoClEta += tc->pt()*tc->eta();
-                protoClPhi += tc->pt()*tc->phi();
-            }
-        }
-        l1t::HGCalCluster cluster( reco::LeafCandidate::LorentzVector(), C2d_hwPtEm + C2d_hwPtHad, 0, 0);
-        cluster.setHwPtEm(C2d_hwPtEm);
-        cluster.setHwPtHad(C2d_hwPtHad);
-
-        if((cluster.hwPtEm()+cluster.hwPtHad())>tc_CUT_){
-            C2d_eta = protoClEta/C2d_pt;
-            C2d_phi = protoClPhi/C2d_pt;                
-            math::PtEtaPhiMLorentzVector calibP4(C2d_pt, C2d_eta, C2d_phi, 0 );
-            cluster.setP4(calibP4);
-            cluster_product_.push_back(0,cluster);
+            } 
+
+            clustersTmp.at( targetClu).addTC( *tc );
+            
         }
-    
-} 
+
+    }
+
+    /* making the collection of clusters */
+    for( unsigned i(0); i<clustersTmp.size(); ++i )
+        clusters_.push_back( 0, clustersTmp.at(i) );
```

Comment:
Is there a reason why you use a temporary vector of clusters and you copy all of them in the end? Which is maybe not the most efficient thing to do.